cpp/include/raft/common/detail/nvtx.cuh

```diff
@@ -0,0 +1,202 @@
+  static inline std::mutex mapMutex;
+  /** saturation */
+  static inline constexpr float S = 0.9f;
+  /** value */
+  static inline constexpr float V = 0.85f;
+  /** golden ratio */
+  static inline constexpr float Phi = 1.61803f;
+  /** inverse golden ratio */
+  static inline constexpr float InvPhi = 1.f / Phi;
+};
+
+// all h, s, v are in range [0, 1]
+// Ref: http://en.wikipedia.org/wiki/HSL_and_HSV#Converting_to_RGB
+inline uint32_t hsv2rgb(float h, float s, float v)
+{
+  uint32_t out = 0xff000000u;
+  if (s <= 0.0f) { return out; }
+  // convert hue from [0, 1] range to [0, 360]
+  float h_deg = h * 360.f;
+  if (0.f > h_deg || h_deg >= 360.f) h_deg = 0.f;
+  h_deg /= 60.f;
+  int h_range = (int)h_deg;
+  float h_mod = h_deg - h_range;
+  float x     = v * (1.f - s);
+  float y     = v * (1.f - (s * h_mod));
+  float z     = v * (1.f - (s * (1.f - h_mod)));
+  float r, g, b;
+  switch (h_range) {
+    case 0:
+      r = v;
+      g = z;
+      b = x;
+      break;
+    case 1:
+      r = y;
+      g = v;
+      b = x;
+      break;
+    case 2:
+      r = x;
+      g = v;
+      b = z;
+      break;
+    case 3:
+      r = x;
+      g = y;
+      b = v;
+      break;
+    case 4:
+      r = z;
+      g = x;
+      b = v;
+      break;
+    case 5:
+    default:
+      r = v;
+      g = x;
+      b = y;
+      break;
+  }
+  out |= (uint32_t(r * 256.f) << 16);
+  out |= (uint32_t(g * 256.f) << 8);
+  out |= uint32_t(b * 256.f);
+  return out;
+}
+
+/**
+ * @brief Helper method to generate 'visually distinct' colors.
+ * Inspired from https://martin.ankerl.com/2009/12/09/how-to-create-random-colors-programmatically/
+ * However, if an associated tag is passed, it will look up in its history for
+ * any generated color against this tag and if found, just returns it, else
+ * generates a new color, assigns a tag to it and stores it for future usage.
+ * Such a thing is very useful for nvtx markers where the ranges associated
+ * with a specific tag should ideally get the same color for the purpose of
+ * visualizing it on nsight-systems timeline.
+ * @param tag look for any previously generated colors with this tag or
+ * associate the currently generated color with it
+ * @return returns 32b RGB integer with alpha channel set of 0xff
+ */
+inline uint32_t generateNextColor(const std::string& tag)
+{
+  // std::unordered_map<std::string, uint32_t> ColorGenState::allColors;
+  // std::mutex ColorGenState::mapMutex;
+
+  std::lock_guard<std::mutex> guard(ColorGenState::mapMutex);
+  if (!tag.empty()) {
+    auto itr = ColorGenState::allColors.find(tag);
+    if (itr != ColorGenState::allColors.end()) { return itr->second; }
+  }
+  float h = rand() * 1.f / RAND_MAX;
+  h += ColorGenState::InvPhi;
+  if (h >= 1.f) h -= 1.f;
+  auto rgb = hsv2rgb(h, ColorGenState::S, ColorGenState::V);
+  if (!tag.empty()) { ColorGenState::allColors[tag] = rgb; }
+  return rgb;
+}
+
+static inline nvtxDomainHandle_t domain = nvtxDomainCreateA("raft");
```

Comment:
Previously in cuml, this code was only used in cuml, so the domain name was fixed. Now, it can be used in raft and in any of its dependencies. But we don't provide a way to set/change the NVTX domain. 

Any suggestions how we could improve that? Preferably, without adding an extra argument to each helper?

Reply:
To avoid overcomplicating this, I just changed the domain name to a more neutral "application".